File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations
def solution(numbers):
    N = len(numbers)
    answers = set()
    MAX = max(numbers)
    is_prime = prime()
    seq = [ permutations(numbers,k) for k in range(1, N+1)]
    seq2 = []
    for a in seq:
        seq2 += a
    for num_str in seq2:
        num = int(''.join(num_str))
        if is_prime[num]:
            answers.add(num)

    return len(answers)

def prime():
    N = 10_000_000
    is_prime = [True]*(N+1)
    is_prime[0], is_prime[1] = False, False
    is_prime[2] = True
    for i in range(2, N+1):
        if not is_prime[i]: continue
        for j in range(2*i, N+1, i):
            is_prime[j] = False
    return is_prime

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.js
let N = 7;
const isPrime = Array.from({length:10**N+1}, ()=>true);
const primes = new Set();
function prime(){
    isPrime[0] = false;
    isPrime[1] = false;
    for (let i=2; i<=10**N;i++) {
        if (!isPrime[i]) continue;
        primes.add(i);
        for (let j=2*i; j<=10**N; j+=i){
            isPrime[j] = false;
        }
    }    
}

const ans = new Set();
const visited = new Set();
const perms = [];
let numbers;
function dfs(i) {
    const num = Number(perms.map(i=>numbers[i]).join(''));
    if (isPrime[num]) ans.add(num)
    if (i === N) {
        return;
    }
    
    for (let j=0; j<N; j++){
        if (visited.has(j)) continue;
        visited.add(j);
        perms.push(j);
        dfs(i+1);
        visited.delete(j)
        perms.pop();
    }
}

function solution(nums) {
    numbers = nums;
    N = nums.length;
    prime();
    dfs(0);
    console.log(ans)
    return ans.size
}
